include/soci/rowset.h

```diff
@@ -41,10 +41,6 @@ class rowset_iterator
     rowset_iterator(statement & st, T & define)
         : st_(&st), define_(&define)
     {
-        assert(0 != st_);
```

Comment:
> rowset_impl::begin() uses `rowset_impl::st_ which is known to never be NULL.
--------------------------------------------------------------------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

That's exactly why the assertion is there.

> IMHO there is no point in keeping this assert as it's tautologically true.

That exactly my point. An assertion in code documents that fact explicitly, so reader does not have to go through the code to come to those 1-2-3 conclusions.

Those particular assertions I'm discussing tell those pointers mustn't be nullptr. It could be a comment, comments come and go, whereas assertion has higher severity.

>  if we think we must guard against the pointer being null/empty here, let's do it -- 
> but just not using assert() but by throwing an exception

I completely agree.
It's just that those assertions don't do that here. They don't guard against nullptr. They just explicitly document those pointers mustn' be null there.

IOW, if `assert(0 != st_);` is not fulfilled and causes a program to abort, then even if there is no assertion here, the program should abort or crash anyway. Throwing exception there would cause invalid questions: so, there can be a `rowset` object without statement attached, can't there?

I'm not arguing to keep those assertions, they can go.
I'm just surprised those assertions are considered harmful to end users (if they can, they will be harmed without those assertions anyway).